src/VisualStudio/Core/Def/Telemetry/VisualStudioWorkspaceTelemetryService.cs

```diff
@@ -44,24 +41,26 @@ protected override ILogger CreateLogger(TelemetrySession telemetrySession, bool
 
     protected override void TelemetrySessionInitialized()
     {
+        var cancellationToken = _threadingContext.DisposalToken;
         _ = Task.Run(async () =>
         {
-            var client = await RemoteHostClient.TryGetClientAsync(_workspace, CancellationToken.None).ConfigureAwait(false);
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
             if (client == null)
-            {
                 return;
-            }
 
             var settings = SerializeCurrentSessionSettings();
             Contract.ThrowIfNull(settings);
 
             // Only log "delta" property for block end events if feature flag is enabled.
             var logDelta = _globalOptions.GetOption(DiagnosticOptionsStorage.LogTelemetryForBackgroundAnalyzerExecution);
 
+            var statusService = _workspace.Services.GetRequiredService<IWorkspaceStatusService>();
+            await statusService.WaitUntilFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
This worries me for two reasons:

1. If this service glitches (which it does sometimes) we'll then never have remote telemetry.
2. We could get remote calls prior to this running, right? So we have no telemetry coverage in this case?

Is there a reason we don't just do this on the first call that's launching OOP, whatever other thing that is?

Reply:
> If this service glitches (which it does sometimes) 

We're screwed if this is *actually* the case.  A ton of features wait on this.  And if it never fires, they all die then.

> We could get remote calls prior to this running, right?  

Does that matter?  We should assume we're never gonig to get 100% of all telemetry we try to store.  We shoudln't depend on that.  It should be best effort and should always be secondary to the user's needs.

Reply:
> Is there a reason we don't just do this on the first call that's launching OOP, whatever other thing that is?

Reply:
> Is there a reason we don't just do this on the first call that's launching OOP, whatever other thing that is?

@tmat  is this possible?

Reply:
> Is there a reason we don't just do this on the first call that's launching OOP, whatever other thing that is?

This is exactly how we already handle `IRemoteProcessTelemetryService.InitializeAsync`. I believe the behavior we want for this method is to invoke `InitializeTelemetrySessionAsync` after _both_ of the following occur:

1. `TelemetrySessionInitialized` is invoked
2. `IRemoteProcessTelemetryService.InitializeAsync` is invoked

Right now we are only gating on the second one.

Reply:
we're gating on both.  IRemoteProcessTelemetryService.InitializeAsync is called in ServiceHubRemoteHostClient.CreateAsync.  Whcih we're gating on, because we're the one calling that with the RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken) code dierectly below this.

So we only initialize after both happen.  But hte problem is that *we're* causing the latter to happen.